apps/emqx_machine/src/emqx_machine.erl

```diff
@@ -0,0 +1,165 @@
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_machine).
+
+-export([start/0,
+         graceful_shutdown/0
+        ]).
+
+-export([ stop_apps/1
+        , ensure_apps_started/0
+        ]).
+
+-export([sorted_reboot_apps/0]).
+
+-include_lib("emqx/include/logger.hrl").
+
+%% @doc EMQ X boot entrypoint.
+start() ->
+    ok = set_backtrace_depth(),
+    ok = print_otp_version_warning(),
+
+    %% need to load some app envs
+    %% TODO delete it once emqx boot does not depend on modules envs
+    _ = load_modules(),
+    ok = load_config_files(),
+
+
+    ok = ensure_apps_started(),
+
+    _ = emqx_plugins:load(),
+
+    ok = print_vsn(),
+
+    ok = start_autocluster(),
+    ok = emqx_machine_terminator:start().
+
+graceful_shutdown() ->
+    emqx_machine_terminator:graceful_wait().
+
+set_backtrace_depth() ->
+    {ok, Depth} = application:get_env(emqx_machine, backtrace_depth),
+    _ = erlang:system_flag(backtrace_depth, Depth),
+    ok.
+
+-if(?OTP_RELEASE > 22).
+print_otp_version_warning() -> ok.
+-else.
+print_otp_version_warning() ->
+    ?ULOG("WARNING: Running on Erlang/OTP version ~p. Recommended: 23~n",
+          [?OTP_RELEASE]).
+-endif. % OTP_RELEASE > 22
+
+-ifdef(TEST).
+print_vsn() -> ok.
+-else. % TEST
+print_vsn() ->
+    ?ULOG("~s ~s is running now!~n", [emqx_app:get_description(), emqx_app:get_release()]).
+-endif. % TEST
+
+-ifndef(EMQX_ENTERPRISE).
+load_modules() ->
+    application:load(emqx_modules).
+-else.
+load_modules() ->
+    ok.
+-endif.
+
+load_config_files() ->
+    %% the app env 'config_files' for 'emqx` app should be set
+    %% in app.time.config by boot script before starting Erlang VM
+    ConfFiles = application:get_env(emqx, config_files, []),
+    %% emqx_machine_schema is a superset of emqx_schema
+    ok = emqx_config:init_load(emqx_machine_schema, ConfFiles),
+    %% to avoid config being loaded again when emqx app starts.
+    ok = emqx_app:set_init_config_load_done().
+
+start_autocluster() ->
+    ekka:callback(prepare, fun ?MODULE:stop_apps/1),
+    ekka:callback(reboot,  fun ?MODULE:ensure_apps_started/0),
+    _ = ekka:autocluster(emqx), %% returns 'ok' or a pid or 'any()' as in spec
+    ok.
+
+stop_apps(Reason) ->
+    ?SLOG(info, #{msg => "stopping_apps", reason => Reason}),
+    _ = emqx_alarm_handler:unload(),
+    lists:foreach(fun stop_one_app/1, lists:reverse(sorted_reboot_apps())).
+
+stop_one_app(App) ->
+    ?SLOG(debug, #{msg => "stopping_app", app => App}),
+    application:stop(App).
+
+ensure_apps_started() ->
+    lists:foreach(fun start_one_app/1, sorted_reboot_apps()).
+
+start_one_app(App) ->
+    ?SLOG(debug, #{msg => "starting_app", app => App}),
+    case application:ensure_all_started(App) of
+        {ok, Apps} ->
+            ?SLOG(debug, #{msg => "started_apps", apps => [App | Apps]});
```

Comment:
[App | lists:delete(App,Apps)].

Reply:
Ah, I thought this list only includes dependent apps.
it actually includes the app itself, will remove `[App |`